src/server/authenticate.rs

```diff
@@ -0,0 +1,177 @@
+                    cert.verify_is_valid_for_dns_name(dns_name)
+                        .map_err(TLSError::WebPKIError)?;
+                }
+
+                // Acquire consensus to trust the root of presented_certs chain
+                let fingerprint = digest::digest(
+                    &digest::SHA1_FOR_LEGACY_USE_ONLY,
+                    &presented_certs[untrusted_index].0
+                );
+                if self.assume_yes {
+                    eprintln!(
+                        "Trusting unknown server certificate: {:?}",
+                        fingerprint,
+                    );
+                } else {
+                    if let Ok(answer) = question::Confirm::new(
+                        format!(
+                            "Unknown server certificate: {:?}. Trust?",
+                            fingerprint,
+                        )
+                    ).default(false).ask() {
+                        if !answer {
+                            return Err(e);
+                        }
+                    } else {
+                        return Err(e);
+                    }
+                }
+
+                // Export the cert in PEM format and return verification success
+                *self.cert_out.lock().unwrap() = Some(
+                    pem::encode(&pem::Pem {
+                        tag: "CERTIFICATE".into(),
+                        contents: presented_certs[untrusted_index].0.clone(),
+                    })
+                );
+            }
+        }
+
+        Ok(ServerCertVerified::assertion())
+    }
+}
+
+pub async fn authenticate(cmd: &Authenticate, opts: &Options) -> anyhow::Result<()> {
+    let builder = opts.conn_params.get()?;
+    let mut creds = builder.as_credentials()?;
+    let verifier = Arc::new(
+        InteractiveCertVerifier::new(
+            cmd.assume_yes,
+            creds.tls_verify_hostname,
+            creds.tls_cert_data.is_none(),
+        )
+    );
+    builder.connect_with_cert_verifier(Some(verifier.clone())).await?;
+    if let Some(cert) = &*verifier.cert_out.lock().unwrap() {
+        creds.tls_cert_data = Some(cert.clone());
+    }
+
+    let cred_path = match &cmd.name {
+        Some(name) => credentials::path(name),
+        None => {
+            if cmd.assume_yes {
+                anyhow::bail!("Instance name required.")
+            }
+            let mut q = question::String::new(
+                "Specify a new instance name for the remote server"
+            );
+            let default = builder.get_addr().to_string().chars().map(|x| match x {
+                'A'..='Z' => x,
+                'a'..='z' => x,
+                '0'..='9' => x,
+                _ => '_',
+            }).collect::<String>();
+            q.default(&default);
+            credentials::path(&q.ask()?)
+        }
+    }?;
+    if cred_path.exists() {
+        if cmd.assume_yes {
+            eprintln!("{} will be overwritten!", cred_path.display());
```

Comment:
I would say `Owerwriting {filename}`. Since "will be" might be interpreted as "would be".